### PR TITLE
[language][vm] vm references & vector specialization

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary
This implements a new design of vm references that offers cleaner and more powerful APIs, which enables specialization of Move values in the VM. Specialization of u64 & bool vectors are also included as an example.

## Test Plan
cargo test

### CODE REVIEW

**Comment by vgao1996** [comment]
@dariorussi FYI I migrated to a new model and updated the PR. It's by no means complete, but should give you a feeling of how the core reference trait & specialized vectors should look like.

File: language/vm/vm-runtime/vm-runtime-types/src/value.rs [hunk 1]
```diff
@@ -131,12 +121,475 @@ pub struct RootAccessPath {
...
+/// behind the shared pointer.
+#[derive(Debug, Clone)]
+pub enum DirectRef {
+    Local(LocalRef),
+    Global(GlobalRef),
+}
+
+impl DirectRef {
+    // TODO: revisit.
+    pub fn from_value_for_cost_synthesis(v: Value) -> Self {
```
Comment (vgao1996) [inline language/vm/vm-runtime/vm-runtime-types/src/value.rs]: @tzakian Does this look good to you? On one hand we have cost_synthesis creating various values, and on the other hand the old philosophy seemed to be that values should be created inside this module and behave as black boxes to the outside world.

**Comment by vgao1996** [comment]
@dariorussi Thank you so much for the review! Those are good points and I think
1) Interpreter working over a common `Reference` type would work. 
  - Yes, totally possible. I'll add a `fn borrow_field` to `Reference` so `DirectRef` isn't exposed to the outside world.
2) `ValueImpl` should be as flat as we can.
  - Doable. The scheme I proposed already does this to some extent.
3) Promotion for a general ValueImpl in Vector or any other "container" is something we don't want to do for now.
  - Yeah, will remove. `NativeVector` will contain `MutVal`s.
4) Abstraction over `Ref` and `RefMut`.
  - I'm not too sure about this part. My intuition is that an alternative implementation without them will be a bit messy with duplicated matches in a few places. But it's a good point and I agree we should at least compare it with the alternatives. I'll play with your playground example and try to implement a set of `write_struct/read_struct` APIs.

**Comment by davidiw** [comment]
Please re-base and re-request if still valid.

**Pull request closed without merging**